### ISSUE TITLE
Node params defined programmatically AFTER app launch: which way?

### ISSUE THREAD

**Issue by juliencarponcy** [timeflux/timeflux#79]
Hi,

I was wondering if you thought of the case where the params for some nodes are received from an external app, and thus after the launch of the app ? 
I guess this situation rules out the config files option.
Would it make more sense (or has always been designed purposefully) to retrieve that information from the app, publish it under a particular zmq topic, and then update some node object attribute ? 
Would overriding an attribute passed as params in the yaml file pause a particular issue (I don't foresee?

I will probably try to play around with these options, but I was curious to have your inputs on that one. 

Thanks very much in advance,

Julien

**Comment by mesca** [comment]
Hi Julien!
The current way to change parameters on the fly (after launching the app) is to handle event messages at the node level. This has been implemented in some (unpublished) nodes, but you're right, a generic solution is needed. This is actually on my todo list, and I expect to have a solution for this by the end of the month.
Do you have any specific use case in mind?

**Comment by juliencarponcy** [comment]
Thanks very much, it make more sense now. I was indeed looking in which nodes it would be best to do that process and how to implement and store them as node attributes. I guess I would then overwrite node._attributes for the parameters to be updated.
The use-case is actually an adptation of your CVEP speller to an external app preseting flickers.
This unity app is so far communicating and setting up the sequences, nb of targets, seq length, nb of cycles etc. in several LSL messages when setting up the different phases of the experiment. 
I parse these messages in a dedicated node and publish them on an event topic, to which the ML nodes and other are connected.

So now I just will make trial&error to see where i can do this process of overwriting attributes. 
The thinking I had now is that I was a bit reluctant to edit timeflux nodes (by opposition to my own plugin nodes) as to be able to keep in sync with future timeflux update. Hence i hesitate to do that either in the ml / epoch node. 

Maybe a good way would be to do that before the epoch node ? Or maybe in a dedicated step of the classification? That's what i'm struggling with right now.

**Comment by mesca** [comment]
Right now, it is not easy to change the parameters of existing nodes after they have been initialized. This is what the next update will be about, but you'll have to wait for a few weeks for it to be ready.

Meanwhile, what you can do is prepare all the needed parameters from Unity, and launch Timeflux only when you're ready using `Process.Start()`. You can pass environment variables using the `-e` flag that will be used by your Timeflux graphs (see: https://doc.timeflux.io/en/stable/usage/getting_started.html#environment). Of course, you'll still have LSL to trigger epochs, etc. Does that make sense?

One small piece of advice: activate vsync on Unity and use the callback to send the LSL messages, that way the epochs timestamps will be synced with the screen refresh.

**Comment by juliencarponcy** [comment]
Yes it does make sense. Indeed my first attempt failed. 
But I am reminded by your answer that the nodes are probably initialized in their own thread and hard to update once created. I guess that would necessitate callbacks and so and that's why it's not as easy as just overriding attributes on update()?

**Comment by mesca** [comment]
Yes, the parameters are passed only once during initialization. Again, I'm working on a mechanism that will allow ro reinit with custom parameters. Although in your case, I still believe that starting a Timeflux process from Unity after you determine your settings is the way to go.

**Comment by juliencarponcy** [comment]
Hi,

I just wanted to know if you were still planning to implement a way to update the params of the node while timeflux is already running?

I did as you suggested with regards to launching timeflux from a third python script upon reception of an LSL marker containing all the necessary values for configuration, writing them at the bottom of the .cnf file, and starting timeflux with that .cnf file programmatically created.

However, we are now foreseeing that this won't work for other features we had in mind. For instance, if we would like to update the predict node threshold according to performance of cVEP classification in a subset of trial. 

Hence my request for update if you're still planning to implement that feature, which i'm sure is not trivial.

Thanks again, 

Julien
